gprofiler/metadata/system_metadata.py

```diff
@@ -0,0 +1,240 @@
+import array
+import fcntl
+import os
+import platform
+import re
+import socket
+import struct
+import subprocess
+import sys
+import time
+from typing import Optional, Tuple
+
+import distro  # type: ignore
+import psutil
+
+from gprofiler.log import get_logger_adapter
+
+logger = get_logger_adapter(__name__)
+hostname: Optional[str] = None
+
+IS_64BIT = sys.maxsize > 2 ** 32
+IFNAMSIZ = 16
+IFF_LOOPBACK = 8
+MAC_BYTES_LEN = 6
+SIZE_OF_SHORT = struct.calcsize('H')
+SIZE_OF_STUCT_ifreq = 40 if IS_64BIT else 32
+MAX_BYTE_COUNT = 1024
+
+
+def get_libc_version() -> Tuple[str, str]:
+    # platform.libc_ver fails for musl, sadly (produces empty results).
+    # so we'll run "ldd --version" and extract the version string from it.
+    # not passing "encoding"/"text" - this runs in a different mount namespace, and Python fails to
+    # load the files it needs for those encodings (getting LookupError: unknown encoding: ascii)
+    def decode_libc_version(version: bytes) -> str:
+        return version.decode("utf-8", errors="replace")
+
+    # Avoid circular imports
+    from gprofiler.utils import run_process
+
+    ldd_version = run_process(
+        ["ldd", "--version"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, suppress_log=True, check=False
+    ).stdout
+    # catches GLIBC & EGLIBC
+    m = re.search(br"GLIBC (.*?)\)", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # catches GNU libc
+    m = re.search(br"\(GNU libc\) (.*?)\n", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # musl
+    m = re.search(br"musl libc.*?\nVersion (.*?)\n", ldd_version, re.M)
+    if m is not None:
+        return "musl", decode_libc_version(m.group(1))
+
+    return "unknown", decode_libc_version(ldd_version)
+
+
+def get_run_mode_and_deployment_type() -> Tuple[str, str]:
+    if os.getenv("GPROFILER_IN_K8S") is not None:  # set in k8s/gprofiler.yaml
+        return "k8s", "k8s"
+    elif os.getenv("GPROFILER_IN_CONTAINER") is not None:  # set by our Dockerfile
+        return "container", "containers"
+    elif os.getenv("STATICX_BUNDLE_DIR") is not None:  # set by staticx
+        return "standalone_executable", "instances"
+    else:
+        return "local_python", "instances"
+
+
+def get_local_ip() -> str:
+    try:
+        local_ips = [ip for ip in socket.gethostbyname_ex(socket.gethostname())[2] if not ip.startswith("127.")]
+    except Exception:
+        local_ips = []
+    if not local_ips:
+        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        try:
+            s.connect(("8.8.8.8", 53))
+            local_ips.append(s.getsockname()[0])
+        finally:
+            s.close()
+    return local_ips[0] if local_ips else "unknown"
+
+
+def get_mac_address() -> str:
```

Comment:
Remind me again why not `getmac`? :sweat_smile: 

Anyway, if keeping it - add an explaining comment about why not using `getmac` and/or why implemented this way (and what is "this way", what's the logic this function is employing with the loopback etc)

Reply:
It's so the MAC is the always the same one as the one fetched by the gAgent. If we used `getmac` here, a different interface could be potentially chosen. Will leave a comment

Reply:
Let's talk about it IRL

Reply:
@Jongy to split it up

Reply:
I decided to skip splitting, it's easier to reuse the same struct objects etc, instead of creating new raw array objects for each call.

Reply:
Added some comments instead